core/src/client.rs

```diff
@@ -421,29 +384,55 @@ impl APIClient {
         Ok(resp)
     }
 
-    pub async fn upload_to_stage(
-        &self,
-        stage_location: &str,
-        data: impl AsyncRead + Send + Sync + 'static,
-        size: u64,
-    ) -> Result<()> {
+    async fn get_presigned_upload_url(&self, stage: &str) -> Result<PresignedResponse> {
+        let sql = format!("PRESIGN UPLOAD {}", stage);
+        let resp = self.query_wait(&sql).await?;
+        if resp.data.len() != 1 {
+            return Err(Error::Request(
+                "Empty response from server for presigned request".to_string(),
+            ));
+        }
+        if resp.data[0].len() != 3 {
+            return Err(Error::Request(
+                "Invalid response from server for presigned request".to_string(),
+            ));
+        }
+        // resp.data[0]: [ "PUT", "{\"host\":\"s3.us-east-2.amazonaws.com\"}", "https://s3.us-east-2.amazonaws.com/query-storage-xxxxx/tnxxxxx/stage/user/xxxx/xxx?" ]
+        let method = resp.data[0][0].clone();
+        if method != "PUT" {
+            return Err(Error::Request(format!(
+                "Invalid method for presigned upload request: {}",
+                method
+            )));
+        }
+        let headers: BTreeMap<String, String> =
+            serde_json::from_str(resp.data[0][1].clone().as_str())?;
+        let url = resp.data[0][2].clone();
+        Ok(PresignedResponse {
+            method,
+            headers,
+            url,
+        })
+    }
+
+    pub async fn upload_to_stage(&self, stage: &str, data: Reader, size: u64) -> Result<()> {
         if self.presigned_url_disabled {
-            self.upload_to_stage_with_stream(stage_location, data, size)
-                .await
+            self.upload_to_stage_with_stream(stage, data, size).await
         } else {
-            self.upload_to_stage_with_presigned(stage_location, data, size)
-                .await
+            let presigned = self.get_presigned_upload_url(stage).await?;
```

Comment:
For the storage that does not support the presign, can we use stream instead?

Reply:
We have a client argument `presigned_url_disabled` in DSN which can do this.

Reply:
BTW, this argument only works for HTTP API. FlightSQL will always do a presigned upload.

Reply:
ok